from launch import LaunchDescription
from launch_ros.actions import Node

def generate_launch_description():
    ld= LaunchDescription()
    
    from launch_ros.actions import Node

    ekf_node = Node(
        package="kalman_filter_app",
        executable="kalman_filter_app",
        output='screen',
        parameters=[
            {"Q_process_noise": [0.1, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                                0.0, 0.1, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                                0.0, 0.0, 0.1, 0.0, 0.0, 0.0, 0.0, 0.0,
                                0.0, 0.0, 0.0, 0.1, 0.0, 0.0, 0.0, 0.0,
                                0.0, 0.0, 0.0, 0.0, 0.1, 0.0, 0.0, 0.0,
                                0.0, 0.0, 0.0, 0.0, 0.0, 0.1, 0.0, 0.0,
                                0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.1, 0.0,
                                0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.1]},
            {"Q_rows": 8},
            {"Q_cols": 8},

            {"R_observation_noise": [0.05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                                    0.0, 0.05, 0.0, 0.0, 0.0, 0.0, 0.0,
                                    0.0, 0.0, 0.05, 0.0, 0.0, 0.0, 0.0,
                                    0.0, 0.0, 0.0, 0.05, 0.0, 0.0, 0.0,
                                    0.0, 0.0, 0.0, 0.0, 0.05, 0.0, 0.0,
                                    0.0, 0.0, 0.0, 0.0, 0.0, 0.05, 0.0,
                                    0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.05]},
            {"R_rows": 7},
            {"R_cols": 7},

            {"X_init": [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]},
            {"X_init_size": 8},

            {"Cov_init": [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                        0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                        0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                        0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0,
                        0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0,
                        0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0,
                        0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0,
                        0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0]},
            {"Cov_init_rows": 8},
            {"Cov_init_cols": 8},
        ]
    )

    
    ld.add_action(ekf_node)
    return ld
    